File: java/StringUtils.java
package com.tinyutils;

/**
 * Immutable string transformation utilities.
 *
 * <p>Every method returns a new {@link String}; no state is modified.
 */
public final class StringUtils {

    private StringUtils() {} // utility class

    /** Regex used by {@link #slugify(String)}. */
    private static final java.util.regex.Pattern NON_WORD =
        java.util.regex.Pattern.compile("[^\\w\\s-]");

    private static final java.util.regex.Pattern SPACES =
        java.util.regex.Pattern.compile("[-\\s]+");

    /**
     * Converts free text into a URL-safe slug.
     *
     * @param text input, may be {@code null}
     * @return slug or empty string
     */
    public static String slugify(String text) {
        if (text == null || text.isEmpty()) return "";
        String s = text.toLowerCase().trim();
        s = NON_WORD.matcher(s).replaceAll("");
        s = SPACES.matcher(s).replaceAll("-");
        return s.replaceAll("^-+|-+$", "");
    }

    /**
     * Truncates text at the nearest word boundary.
     *
     * @param text     input
     * @param maxLen   hard character limit
     * @param suffix   appended when truncated
     * @return original or truncated string
     */
    public static String truncate(String text, int maxLen, String suffix) {
        if (text == null || text.length() <= maxLen) return text;
        int space = text.substring(0, maxLen).lastIndexOf(' ');
        return (space > 0 ? text.substring(0, space) : text.substring(0, maxLen)) + suffix;
    }

    public static String truncate(String text, int maxLen) {
        return truncate(text, maxLen, "...");
    }
}

File: dart/string_utils.dart
/// String manipulation utilities.
///
/// Every function is a top-level pure function — no state, no side effects.
library;

/// URL‑safe slug from free-form text.
String slugify(String text) {
  final s = text.toLowerCase().trim();
  return s
      .replaceAll(RegExp(r'[^\w\s-]'), '')
      .replaceAll(RegExp(r'[-\s]+'), '-')
      .replaceAll(RegExp(r'^-+|-+\$'), '');
}

/// Truncate text at the nearest word boundary.
String truncate(String text, {int maxLen = 80, String suffix = '...'}) {
  if (text.length <= maxLen) return text;
  final truncated = text.substring(0, maxLen);
  final lastSpace = truncated.lastIndexOf(' ');
  return (lastSpace > 0 ? truncated.substring(0, lastSpace) : truncated) + suffix;
}

/// Count occurrences of a substring.
int countOccurrences(String text, String pattern) {
  int count = 0;
  int i = 0;
  while ((i = text.indexOf(pattern, i)) != -1) {
    count++;
    i += pattern.length;
  }
  return count;
}
